src/msssim.jl

```diff
@@ -0,0 +1,143 @@
+        (num_scales ∈ 1:length(W)) || throw(ArgumentError("`num_scales` $(num_scales) should be positive integer between [1, $(length(W))]"))
+        all(x-> x>=0, flatten(W)) || throw(ArgumentError("α, β, γ should be non-negative for all scales, instead it's $(W)"))
+        sum(flatten(W)) == 0 && throw(ArgumentError("MS-SSIM must have at least one weight > 0"))
+
+        kernel = centered(kernel)
+        if num_scales ≠ length(W)
+            W ≠ MSSSIM_W && @warn "truncate MS-SSIM weights to scale $num_scales"
+        end
+        W = W[1:num_scales]
+        sw = reduce((w1, w2) -> w1.+w2, W[1:num_scales])
+        if !all(x -> x≈1.0, sw)
+            W ≠ MSSSIM_W[1:num_scales] && @warn "normalize MS-SSIM weights so that (∑α, ∑β, ∑γ) == (1.0, 1.0, 1.0)"
+            W = map(w->w./sw, W)
+        end
+        
+        new{typeof(kernel), Int(num_scales)}(kernel, W)
+    end
+end
+# Weights for α, β, γ in [1]
+const MSSSIM_W = (
+    (0.0448, 0.0448, 0.0448), # α₁, β₁, γ₁
+    (0.2856, 0.2856, 0.2856), # α₂, β₂, γ₂
+    (0.3001, 0.3001, 0.3001), # α₃, β₃, γ₃
+    (0.2363, 0.2363, 0.2363), # α₄, β₄, γ₄
+    (0.1333, 0.1333, 0.1333), # α₅, β₅, γ₅
+)
+# shorthand for αᵢ=βᵢ=γᵢ for all scales i
+function MSSSIM(kernel, W::NTuple{N, <:Real}; num_scales=length(W)) where N
+    MSSSIM(kernel, map(x->(x, x, x), W); num_scales=num_scales)
+end
+
+Base.:(==)(ia::MSSSIM, ib::MSSSIM) = ia.kernel == ib.kernel && ia.W == ib.W
+
+
+# SSIM does not allow for user specifying peakval and K, so we don't allow it here either
+(iqi::MSSSIM)(x, ref) = _msssim(iqi, x, ref)
+assess_msssim(x, ref; num_scales=5) = MSSSIM(num_scales=num_scales)(x, ref)
+
+# Implementation details
+function _msssim(iqi::MSSSIM,
+                 x::Union{GenericGrayImage, AbstractArray{<:AbstractRGB}},
+                 ref::Union{GenericGrayImage, AbstractArray{<:AbstractRGB}};
+                 peakval=1.0,
+                 K=SSIM_K)
+    if size(x) ≠ size(ref)
+        err = ArgumentError("images should be the same size, instead they're $(size(x))-$(size(ref))")
+        throw(err)
+    end
+    if axes(x) ≠ axes(ref)
+        x = OffsetArrays.no_offset_view(x)
+        ref = OffsetArrays.no_offset_view(ref)
+    end
+    C₁, C₂ = @. (peakval * K)^2
+    C₃ = C₂/2
+
+    num_scales = length(iqi.W)
+    T = promote_type(float(eltype(ref)), float(eltype(x)))
+    x = of_eltype(T, x)
+    ref = of_eltype(T, ref)
+
+    mean_lcs = NTuple{3, Float64}[]
+    for i in 1:num_scales
+        lcs = __ssim_map_general(x, ref, iqi.kernel, C₁, C₂, C₃; crop=true)
+        w = iqi.W[i]
+        if w[2] ≈ w[3]
+            # only to mimic the original implementation and compute mean(c .* s)
+            # in this case, MS-SSIM with scale 1 is equivalent to SSIM
+            lcs = (mean(lcs[1]), 1, mean(lcs[2] .* lcs[3]))
+        else
+            # here we compute mean(c) * mean(s)
+            lcs = mean.(lcs)
+        end
```

Comment:
This is quite tricky: the original implementation uses `mean(c .* s)` because `cs` computed more efficiently.
The tradeoff is that this way it only supports cases that `α=β=γ`.

Given that we have exposes `α, β, γ` to users in both `SSIM` and `MS-SSIM`, I now think I'd better to just use `lcs = mean.(lcs)` here.